stream_alert/shared/lookup_tables/driver_s3.py

```diff
@@ -120,6 +144,131 @@ def _reload(self):
+        # This will fall back and use the original data if decompression fails
+        if self._compression:
+            data = Compression.gz_decompress(self, data)
+        else:
+            LOGGER.debug('LookupTable (%s): File does not need decompression')
+
+        # Decode the data; right now we make the assumption that the data is always encoded
+        # as JSON.
+        data = Encoding.json_decode(self, data)
+
+        self._cache.setall(data, self._cache_refresh_minutes)
+        self._cache.set(self._MAGIC_CACHE_TTL_KEY, True, self._cache_refresh_minutes)
+        LOGGER.info('LookupTable (%s): Successfully loaded', self.id)
+
+
+class Compression:
+
+    @staticmethod
+    def gz_decompress(driver, data):
+        try:
+            data = zlib.decompress(data, 47)
+            LOGGER.debug(
+                'LookupTable (%s): Object decompressed to %d byte payload',
+                driver.id,
+                sys.getsizeof(data)
+            )
+        except zlib.error:
+            LOGGER.warning(
+                'LookupTable (%s): Data is not compressed; defaulting to original payload',
+                driver.id
+            )
+        return data
+
+    @staticmethod
+    def gz_compress(driver, data):
+        try:
+            original_size = sys.getsizeof(data)
+            data = zlib.compress(data, level=zlib.Z_BEST_COMPRESSION)
+            LOGGER.debug(
+                'LookupTable (%s): Successfully compressed input data from %d to %d bytes',
+                driver.id,
+                original_size,
+                sys.getsizeof(data)
+            )
+            return data
+        except zlib.error:
+            LOGGER.exception('LookupTable (%s): Data compression error.', driver.id)
+
+
+class Encoding:
+    """
+    Encapsulation of encoding algorithms for S3 data.
+
+    Right now we only support JSON encoding. In the future, we could potentially support
+    """
+
+    @staticmethod
+    def json_encode(driver, data):
+        try:
+            return json.dumps(data)
+        except (ValueError, TypeError):
+            LOGGER.exception(
+                'LookupTable (%s): Failed to json encode data', driver.id
+            )
+
+    @staticmethod
+    def json_decode(driver, data):
+        try:
+            data = json.loads(data)
+            LOGGER.debug(
+                'LookupTable (%s): File successfully JSON decoded. Discovered %s keys.',
+                driver.id,
+                len(data)
+            )
+            return data
+        except ValueError:
+            LOGGER.exception(
+                'LookupTable (%s): Failed to json decode data', driver.id
+            )
+
+
+class S3Adapter:
+    """Adapter class that manages uploading data to and downloading data from AWS S3"""
+
+    def __init__(self, driver, boto_s3_client, s3_bucket, s3_key):
+        self._driver = driver
+        self._s3_client = boto_s3_client
```

Comment:
if I'm reading this right, the `boto_s3_client` here is actually an `S3.ServiceResource`. instead of storing this, couldn't you store the `Bucket` resource directly (which is created/used in the `upload` method)?

Reply:
I'm using the `Object` resource in the `download` method so this wouldn't work. I could refactor it to use `Bucket` and `download_fileobj` but I wanted the code to be as similar to the original LookupTables code (.. where possible) to reduce chance of scalability issues

Reply:
the `Bucket` resource has an `Object` method that returns an `Object` resource: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Bucket.Object